cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,458 @@
+    //                                       stream_);
+    
+    // if (lcol.has_nulls() || rcol.has_nulls())
+    //   {
+    //     auto merged_view = column->mutable_view();
+    //     materialize_bitmask(lcol,
+    //                         rcol,
+    //                         merged_view,
+    //                         dv_row_order_.data().get(),
+    //                         stream_);
+    //   }
+    // return column;
+  }
+
+private:
+  index_vector const& dv_row_order_;
+  rmm::mr::device_memory_resource* mr_;
+  cudaStream_t stream_;
+};
+  
+
+  std::unique_ptr<cudf::experimental::table> merge(cudf::table_view const& left_table,
+                                                   cudf::table_view const& right_table,
+                                                   std::vector<cudf::size_type> const& key_cols,
+                                                   std::vector<cudf::order> const& column_order,
+                                                   std::vector<cudf::null_order> const& null_precedence,
+                                                   rmm::mr::device_memory_resource* mr,
+                                                   cudaStream_t stream = nullptr) {
+    auto n_cols = left_table.num_columns();
+    CUDF_EXPECTS( n_cols == right_table.num_columns(), "Mismatched number of columns");
+    if (left_table.num_columns() == 0) {
+      return cudf::experimental::empty_like(left_table);
+    }
+
+    CUDF_EXPECTS(cudf::have_same_types(left_table, right_table), "Mismatched column types");
+
+    auto keys_sz = key_cols.size(); 
+    CUDF_EXPECTS( keys_sz > 0, "Empty key_cols");
+    CUDF_EXPECTS( keys_sz <= static_cast<size_t>(left_table.num_columns()), "Too many values in key_cols");
+
+    CUDF_EXPECTS(keys_sz == column_order.size(), "Mismatched size between key_cols and column_order");
+    
+    if (not column_order.empty())
+      {
+        CUDF_EXPECTS(column_order.size() <= static_cast<size_t>(left_table.num_columns()), "Too many values in column_order");
+      }
+
+    //collect index columns for lhs, rhs, resp.
+    //
+    cudf::table_view index_left_view{left_table.select(key_cols)};
+    cudf::table_view index_right_view{right_table.select(key_cols)};
+    bool nullable = cudf::has_nulls(index_left_view) || cudf::has_nulls(index_right_view);
+
+    //extract merged row order according to indices:
+    //
+    rmm::device_vector<index_type>
+      merged_indices = generate_merged_indices(index_left_view,
+                                               index_right_view,
+                                               column_order,
+                                               null_precedence,
+                                               nullable);
+
+    //create merged table:
+    //
+    std::vector<std::unique_ptr<column>> v_merged_cols;
+    v_merged_cols.reserve(n_cols);
+
+    column_merger merger{merged_indices, mr, stream};
+    
+    for(auto i=0;i<n_cols;++i)
```

Comment:
std::transform

Reply:
That won't work with `emplace_back()`. `std::transform` will need a lambda to return the data to be filled. But that will be by copy, and it wouldn't take advantage of the benefits of `emplace_back`.

Notice, here, the vector is populated. It would need an equivalent of `std::back_inserter` adaptor, but working with `emplace`.